### PR TITLE
Virtual library implementation fixes related to private modules

### PR DESCRIPTION
**Pull request by rgrinberg**
Implementations of virtual libraries need an alias module that includes modules from the implementation and virtual library. On the way, I also fix two unrelated issues:

* The dependency graph was wrong in the presence of private module mli's
* The .install file wasn't being generated as some incorrect copy rules were setup.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Okay, so this PR fixes all the glaring bugs in virtual libraries. The next PR should clean up the mess.

There's still a couple of edge cases to test for:
* That implementations do not introduce lib interface modules
* That private modules in implementations and virtual libs can happily co-exist.

File: src/lib_modules.ml [hunk 1]
```diff
@@ -50,9 +50,18 @@ let make_wrapped ~(lib : Dune_file.Library.t) ~dir ~transition ~modules
...
       wrap_modules modules, Module.Name.Map.empty
   in
+  let implements = Dune_file.Library.is_impl lib in
   let alias_module =
     let alias_prefix =
-      String.uncapitalize (Module.Name.to_string main_module_name) in
+      let base =
+        String.uncapitalize (Module.Name.to_string main_module_name) in
+      if implements then
+        sprintf "%s_%s__" base (Lib_name.Local.to_string (snd lib.name))
```
Comment (diml) [inline src/lib_modules.ml]: Should we use `"%s__%s__`? To avoid clashes between implementation `bar` of library `foo` and library `foo_bar`

File: src/lib_modules.ml [hunk 2]
```diff
@@ -63,14 +72,14 @@ let make_wrapped ~(lib : Dune_file.Library.t) ~dir ~transition ~modules
 
          https://github.com/ocaml/dune/issues/567 *)
       Some
-        (Module.make (Module.Name.add_suffix main_module_name "__")
+        (Module.make (Module.Name.add_suffix alias_module_name "__")
```
Comment (diml) [inline src/lib_modules.ml]: Don't we end up adding 4 trailing underscores here?

Comment (diml) [reply src/lib_modules.ml]: Looking at the tests, it seems that we don't. However, reading the code, it really seems to me that we should end up with 4 underscores. @rgrinberg do you agree?

Comment (rgrinberg) [reply src/lib_modules.ml]: It's because `Module.Name.Map.mem modules main_module_name` is never true i suppose. It could be true in the case where the lib interface module is virtual however. I will fix this.

File: src/lib_modules.ml (hunk 1 above)
Comment (rgrinberg) [reply src/lib_modules.ml]: Yeah, that makes sense. Although I'm wondering if this will cause issues if we introduce more than 1 layer of aliasing. For example, `Foo.Bar.` We can fix that later however.

File: src/lib_modules.ml (hunk 2 above)
Comment (rgrinberg) [reply src/lib_modules.ml]: Now the implementation case is handled separately and the separator is always `__`

File: src/ocamldep.ml [hunk 3]
```diff
@@ -99,18 +99,49 @@ module Dep_graphs = struct
...
-            )
+        { Dep_graph.
+          dir = impl.intf.dir
+        ; per_module =
+            Module.Name.Map.merge vlib.intf.per_module impl.intf.per_module
+              ~f:(fun name vlib impl ->
+                match vlib, impl with
+                | None, None -> assert false
+                | Some v, Some i ->
+                  (* This can happen when we are implementing the virtul module
```
Comment (emillon) [inline src/ocamldep.ml]: typo: virtual

File: src/ocamldep.ml [hunk 4]
```diff
@@ -99,18 +99,49 @@ module Dep_graphs = struct
...
+                     or this is a private module. In the former case, we should
+                     simply use [v], and in the latter, we should use [i]. But
+                     since we can't the tell the cases apart here, we'll just
+                     return the non empty list *)
+                  v &&& i >>^ (function
+                    | [], modules
+                    | modules, [] -> modules
+                    | (_ :: _ as vlib), (_ :: _ as impl) ->
```
Comment (emillon) [inline src/ocamldep.ml]: the _::_ patterns shouldn't be required here. If you're keeping them for documentation it's OK, otherwise that can be simplified to `(vlib, impl)`

File: src/lib_modules.ml (hunk 2 above)
Comment (diml) [reply src/lib_modules.ml]: Thanks, that looks clearer.

File: src/ocamldep.ml (hunk 4 above)
Comment (rgrinberg) [reply src/ocamldep.ml]: Yup, it's only for documentation.

File: src/ocamldep.ml (hunk 3 above)
Comment (rgrinberg) [reply src/ocamldep.ml]: typo is fixed.
